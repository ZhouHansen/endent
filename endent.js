import dedent from 'dedent'
import objectorarray from 'objectorarray'
import parse from 'fast-json-parse'

export default endent

const ENDENT_ID = 'twhZNwxI1aFG3r4'

function endent (strings, ...values) {
  const raw = typeof strings === 'string' ? [strings] : strings.raw

  let result = ''

  for (let i = 0; i < raw.length; i++) {
    result += raw[i]

    if (i < values.length) {
      var value = values[i]
      var isJson = false

      if (parse(value).value) {
        value = parse(value).value
        isJson = true
      }

      if ((value && value[ENDENT_ID]) || isJson) {
        var rawlines = result.split('\n')
        var l = rawlines[rawlines.length - 1].search(/\S/)
        var endentation = ' '.repeat(l)
        var valueJson = isJson ? JSON.stringify(value, null, 2) : value[ENDENT_ID]
        var valueLines = valueJson.split('\n')

        valueLines.forEach((l, index) => {
          if (index > 0) {
            result += ('\n' + endentation + l)
          } else {
            result += l
          }
        })
      } else if (typeof value === 'string' && value.includes('\n')) {
          var endentations = result.match(/(?:^|\n)( *)$/)

          if (endentations && typeof value === 'string') {
            var endentation = endentations[1]
            result += value
              .split('\n')
              .map((str, i) => i === 0 ? str : `${endentation}${str}`)
              .join('\n')
            } else {
              result += value
            }
      } else {
        result += value
      }
    }
  }

  return dedent(result)
}

endent.pretty = (data) => {
  return objectorarray(data) ? {[ENDENT_ID]: JSON.stringify(data, null, 2)} : data
}
