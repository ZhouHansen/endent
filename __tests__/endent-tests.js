import endent from '../endent.js'

describe('endent', () => {
  it('only endent object or array', () => {
    var json = JSON.stringify(JSON.parse('[ "abc" ]'), null, 2)

    var someobj = {
      contact: {
        jack: '123456',
        tom: '654321'
      },
      name: 'template',
      color: 'blue',
      animals: [ 'bear', 'fish', 'dog', 'cat']
    }

    var colors = ['red', 'pink', 'white']
    var objectName = "someobj"

    var dependencies = ['jquery', 'underscore', 'bootstrap']
    var dependencyTmpl = ``

    dependencies.forEach((d, i) => {
      dependencyTmpl += `var ${d} = require("${d}")\n`
    })

    var jsFile = endent`
      ${dependencyTmpl}
      module.exports = store

      function store (state, emitter) {
        emitter.on("DOMContentLoaded", function () {
          state["json"] = ${json}
          state["${objectName}"] = ${endent.pretty(someobj)}
          state["colors"] = ${endent.pretty(colors)}
          state["name"] = "${endent.pretty("jack")}"
          state["name2"] = "${"tom"}"
          state["number"] = ${endent.pretty(123)}
          state["number2"] = ${123}
          state["Iamundefined"] = ${endent.pretty()}
          state["Iamnull"] = ${endent.pretty(null)}
          state["Iamregexp"] = ${endent.pretty(/abc/)}
        })
      }
    `

    expect(jsFile).toMatchSnapshot()
  })
})
