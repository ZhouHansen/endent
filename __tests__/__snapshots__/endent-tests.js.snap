exports[`endent only endent object or array 1`] = `
"var jquery = require("jquery")
var underscore = require("underscore")
var bootstrap = require("bootstrap")

module.exports = store

function store (state, emitter) {
  emitter.on(\"DOMContentLoaded\", function () {
    state[\"json\"] = [
      \"abc\"
    ]
    state[\"someobj\"] = {
      \"contact\": {
        \"jack\": \"123456\",
        \"tom\": \"654321\"
      },
      \"name\": \"template\",
      \"color\": \"blue\",
      \"animals\": [
        \"bear\",
        \"fish\",
        \"dog\",
        \"cat\"
      ]
    }
    state[\"colors\"] = [
      \"red\",
      \"pink\",
      \"white\"
    ]
    state[\"name\"] = \"jack\"
    state[\"name2\"] = \"tom\"
    state[\"number\"] = 123
    state[\"number2\"] = 123
    state[\"Iamundefined\"] = undefined
    state[\"Iamnull\"] = null
    state[\"Iamregexp\"] = /abc/
  })
}"
`;
